xmodule/services.py

```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+
+    def closed(self):
+        """
+        Is the student still allowed to submit answers?
+        """
+        if not self._xblock:
+            return False
+
+        if self.used_all_attempts():
+            return True
+        # Using XBlock's is_past_due method, as it's suited for checking the
+        # submission deadline, unlike the is_past_due method of this service.
+        elif self._xblock.is_past_due():
+            return True
+        return False
+
+    def answer_available(self):
+        """
+        Returns whether correctness is available now, for the given attributes.
+        """
+        if not self._xblock:
+            return False
+
+        show_answer = self._xblock.showanswer
+        max_attempts = self._xblock.max_attempts
+        attempts = self._xblock.attempts
+        is_correct = self._xblock.is_correct()
+        required_attempts = self._xblock.attempts_before_showanswer_button
+        has_due_date = getattr(self._xblock, 'close_date', None) is not None
+        past_due = self.is_past_due()
+        is_attempted = self.is_attempted()
+        used_all_attempts = self.used_all_attempts()
+        closed = self.closed()
+
+        if not self.correctness_available():
+            # If correctness is being withheld, then don't show answers either.
+            return False
+        elif show_answer == ShowAnswer.NEVER:
+            return False
+        elif self._user_is_staff:
+            # This is after the 'never' check because course staff can see correctness
+            # unless the sequence/problem explicitly prevents it
+            return True
+        elif show_answer == ShowAnswer.ATTEMPTED:
+            return is_attempted or (has_due_date and past_due)
+        elif show_answer == ShowAnswer.ANSWERED:
+            # NOTE: this is slightly different from 'attempted' -- resetting the problems
+            # makes lcp.done False, but leaves attempts unchanged.
+            return is_correct
+        elif show_answer == ShowAnswer.CLOSED:
+            return closed
+        elif show_answer == ShowAnswer.FINISHED:
+            return closed or is_correct
+        elif show_answer == ShowAnswer.CORRECT_OR_PAST_DUE:
+            return is_correct or (has_due_date and past_due)
+        elif show_answer == ShowAnswer.PAST_DUE:
+            return past_due
+        elif show_answer == ShowAnswer.AFTER_SOME_NUMBER_OF_ATTEMPTS:
+            if max_attempts and required_attempts >= max_attempts:
+                required_attempts = max_attempts
+            return attempts >= required_attempts
+        elif show_answer == ShowAnswer.ALWAYS:
+            return True
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS:
+            return used_all_attempts
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS_OR_CORRECT:
+            return used_all_attempts or is_correct
+        elif show_answer == ShowAnswer.ATTEMPTED_NO_PAST_DUE:
+            return is_attempted
```

Comment:
Suggestion: I think we could use [the `match` statement](https://peps.python.org/pep-0636/) to make this code more readable.